### PR TITLE
HP-2008: processing insufficient funds

### PR DESCRIPTION
**Pull request by tafid**
## Summary by CodeRabbit

- **New Features**
	- Introduced a new response type for payment failures related to insufficient funds, enhancing the payment handling process.
	- Added a new class to manage insufficient funds responses, improving clarity and error reporting during transactions.
	- Implemented a new exception type for insufficient funds, streamlining error handling for declined payments.

- **Bug Fixes**
	- Improved error handling for payment declines, ensuring users receive more accurate feedback on transaction issues.

### CODE REVIEW

File: src/merchants/stripe/StripeMerchant.php [hunk 1]
```diff
@@ -173,8 +174,31 @@ public function completePurchase($data)
...
 
-        if (isset($response->getData()['error']['message'])) {
-            throw new MerchantException('Failed to charge card: ' . $response->getData()['error']['message']);
+        if (
+            isset($response->getData()['last_payment_error'])
+            && $response->getData()['last_payment_error']['code'] === 'card_declined'
+            && $response->getData()['last_payment_error']['decline_code'] === 'insufficient_funds'
+        ) {
+            return (new InsufficientFundsResponse())
+                ->setIsSuccessful(true)
```
[the remark this answers is not in the document]
Comment (SilverFire) [reply src/merchants/stripe/StripeMerchant.php]: Agree with @coderabbitai

Why do you set it to true?

Comment (hiqsol) [inline src/merchants/stripe/StripeMerchant.php]: why successful for failed transaction?
```suggestion
                ->setIsSuccessful(false)
```

File: src/merchants/stripe/StripeMerchant.php [hunk 2]
```diff
@@ -173,8 +174,31 @@ public function completePurchase($data)
...
+                ->setAmount(new Money($response->getData()['amount'], new Currency(strtoupper($response->getData()['currency']))))
+                ->setFee(new Money(0, new Currency(strtoupper($response->getData()['currency']))))
+                ->setTransactionReference(
+                    $response->getData()['charges']['data'][0]['payment_intent']
+                )
+                ->setTransactionId($response->getTransactionId())
+                ->setPayer(
+                    $response->getCustomerReference()
+                    ?? $response->getData()['charges']['data'][0]['customer']
+                    ?? ''
```
[the remark this answers is not in the document]
Comment (hiqsol) [reply src/merchants/stripe/StripeMerchant.php]: діло каже :)

File: src/merchants/stripe/StripeMerchant.php [hunk 3]
```diff
@@ -25,6 +26,7 @@
 use hiqdev\php\merchant\merchants\RemoteCustomerAwareMerchant;
 use hiqdev\php\merchant\response\CardAuthorizationResponse;
 use hiqdev\php\merchant\response\CompletePurchaseResponse;
+use hiqdev\php\merchant\response\InsufficientFundsResponse;
```
Comment (SilverFire) [inline src/merchants/stripe/StripeMerchant.php]: del?

**Review by SilverFire — APPROVED** [review]

File: src/merchants/stripe/StripeMerchant.php [hunk 4]
```diff
@@ -173,8 +175,17 @@ public function completePurchase($data)
...
         }
 
-        if (isset($response->getData()['error']['message'])) {
-            throw new MerchantException('Failed to charge card: ' . $response->getData()['error']['message']);
+        if (
+            isset($response->getData()['last_payment_error'])
+            && $response->getData()['last_payment_error']['code'] === 'card_declined'
+            && $response->getData()['last_payment_error']['decline_code'] === 'insufficient_funds'
+        ) {
+            throw new InsufficientFundsException($response->getData()['last_payment_error']['message']);
```
Comment (SilverFire) [inline src/merchants/stripe/StripeMerchant.php]: Please add some context to the exception, e.g. whole `getData()`

**Review by SilverFire — APPROVED** [review]

**Pull request merged**